File: patterns/two_pointers/pointer_for_everyone/pattern.py
# Time: O(a+b)
# Memory: O(min(a,b))


def pointer_for_everyone(a: list, b: list) -> list:
    # =============================================================================
    # 1) Pointers initializations
    # =============================================================================
    p1: int = 0
    p2: int = 0
    result: list = []

    # =============================================================================
    # 2) While loop while we check all the conditions
    # =============================================================================
    while p1 < len(a) and p2 < len(b):
        # =============================================================================
        # 3) Pointer logic movement: IS THE KEY TO THE SOLUTION, UNDERSTAND WHEN
        # TO MOVE THEM, is this is understood, the problem is solved automatically
        # =============================================================================
        ...
        # <The L and R pointers are moved here>

    return result

File: patterns/dots_and_segments/two_pointers_on_segments/pattern.cpp
// This has the same approach like in the pattern of a pointer for everyone
//
// Time: O(n+m)
// Memory: O(max(n, m))

#include <vector>

using namespace std;

vector<vector<int>> multi_segment_intersect(const vector<vector<int>> &seg1, const vector<vector<int>> &seg2)
{
    // ============================================================================
    // Stage 1: Initialization
    // ============================================================================
    vector<vector<int>> result;
    int p1 = 0, p2 = 0;

    // ============================================================================
    // Stage 2: Main loop
    // ============================================================================
    while (p1 < seg1.size() && p2 < seg2.size())
    {
        // Stage 3: Logic for the moving pointers and formatting of the result
        // MAIN PROBLEM IS TO UNDERSTAND WHEN TO MOVE P1 OR P2 AND HOW TO FORM THE RESPONSE
        // ...
    }

    return result;
}